keras_core/distribution/distribution_lib.py

```diff
@@ -184,26 +187,108 @@ def get_data_layout(self, data_shape):
-        pass
+        original_scope = distribution()
+        set_distribution(self)
+        try:
+            yield
+        finally:
+            set_distribution(original_scope)
+
+    @property
+    def device_mesh(self):
+        return self._device_mesh
+
+
+class DataParallel(Distribution):
+    def __init__(self, device_mesh=None, devices=None):
+        """Create the data parallel distribution.
+
+        You can choose to create this instance by either specifing
+        the `device_mesh` or `devices` parameters (but not both).
+
+        The device_mesh is expected to be a `DeviceMesh` instance, and is
+        expected to be 1D only. In case that the mesh has multiple axes, then
+        the first axis will be treated as the data parallel dimension
+        (and a warning will be raised).
+
+        When a list of `devices` are provided, they will be used to construct a
+        1D mesh.
+
+        When both `mesh` and `devices` are absent, then `list_devices()`
+        will be used to detect any available devices and create a 1D mesh from
+        them.
+        """
+        if device_mesh:
+            self._initialize_with_device_mesh(device_mesh)
+        elif devices:
+            self._initialize_mesh_from_devices(devices)
+        else:
+            self._initialize_mesh_from_list_devices()
+
+        self._batch_dim_name = self.device_mesh.axis_names[0]
+
+    def _initialize_with_device_mesh(self, device_mesh):
+        if not isinstance(device_mesh, DeviceMesh):
+            raise ValueError(
+                "Expect `mesh` to be an instance of `DeviceMesh`. "
+                f"Received: mesh={device_mesh} (of type {type(device_mesh)})"
+            )
+        super().__init__(device_mesh)
+        if self.device_mesh.devices.ndim != 1:
+            logging.warning(
+                "Expect the input mesh to be 1D, but received "
+                "mesh.devices.ndim=%d. "
+                "The first axis will be used for data-parallel sharding.",
+                device_mesh.devices.ndim,
+            )
+
+    def _initialize_mesh_from_devices(self, devices):
+        devices = np.array(devices)
+        device_mesh = DeviceMesh(
+            shape=devices.shape,
+            axis_names=[DEFAULT_BATCH_DIM_NAME],
+            devices=devices,
+        )
+        super().__init__(device_mesh)
+
+    def _initialize_mesh_from_list_devices(self):
+        devices = np.array(list_devices())
+        device_mesh = DeviceMesh(
+            shape=devices.shape,
+            axis_names=[DEFAULT_BATCH_DIM_NAME],
+            devices=devices,
+        )
+        super().__init__(device_mesh)
+
+    def get_data_layout(self, data_shape):
+        data_shard_spec = [None] * len(data_shape)
+        data_shard_spec[0] = self._batch_dim_name  # Shard on the first dim
+        return TensorLayout(data_shard_spec, self.device_mesh)
+
+    def get_variable_layout(self, variable_shape, variable_path):
```

Comment:
Is the reason why we need both arguments that `variable_path` could map to multiple variables, and they might not have the same shape? Would it be better to instead ensure 1:1 mapping between variables and paths, so that we can refer to a variable by the path only (and then retrieve the shape internally)?

Reply:
The reason is that in the data parallel distribution, since there isn't a layout map, we can't get any layout info from the path. With the variable shape, we can probably create the replicated variables (with X number of None where X is the rank).

Reply:
This is also true for the model parallel distribution for that any variable that doesn't have layout rule specified.

Reply:
Ok, how about we just do `get_variable_layout(variable)`? Then we have access to both `var.shape` and `var.path`. The caller code always has access to the variable instance as far as I can tell.

Reply:
Had a try for this and it failed. The variable.shape is not available at the time when variable._initialize() is call, in fact it is available during the _initialize() and populated right after in https://github.com/keras-team/keras-core/blob/967145263c172dfed9f24ef13d9b655d97a904ea/keras_core/backend/common/variables.py#L72. I was thinking of moving the `variable._shape` to be above the `_initalize()`, but it might have unnecessary complexity that the value is not a standard backend type yet (eg still raw list/int/float before jax.array/tf.convert_to_tensor).

For now I will keep it as is, please let me know if you have any suggestion to change to the base Variable.

Reply:
> I was thinking of moving the variable._shape to be above the _initalize()

I think that should be doable -- if the initializer is a constant, we can call convert_to_tensor on it eagerly when we receive it, and if it's a callable, then we shape must be known in advance and is already set as `._shape` before `_initialize()`.

So I think you can remove line 72 `self._shape = tuple(self._value.shape)` and then above after line 68 `value = initializer(shape, dtype=dtype)` you can add a new line, `self._shape = tuple(value.shape)`.

Reply:
Done. Added a convert_to_tensor, and set the shape before variable._initialize()

Reply:
Well, hit some error for the `convert_to_tensor` approach. Somehow there was a mixture of tf.Variable in the jax/numpy/torch backend for certain layer, and it cause failure when I call backend.convert_to_tensor.

I have changed the logic to set the shape for jax variable in `_initialize` right before it call distribution. This will minimize the change we have to the base variable.